### PR TITLE
Temporary code review for the rate limiting

### PR DESCRIPTION
**Pull request by mangchiandjjoe**
This is not an official code review. Not to be merged.
This PR includes the new rate limiting feature. Some configurations and logic were hard coded.

@qiwzhang

### CODE REVIEW

**Comment by mangchiandjjoe** [comment]
I signed it!

File: src/service_control_client_impl.cc [hunk 1]
```diff
@@ -188,6 +226,83 @@ Status ServiceControlClientImpl::Check(const CheckRequest& check_request,
...
+    // Makes a copy of check_request so that on_done() callback can use
+    // it to call CacheResponse.
+    ::google::api::servicecontrol::v1::AllocateQuotaRequest*
+        quota_request_copy =
+            new ::google::api::servicecontrol::v1::AllocateQuotaRequest(
+                quota_request);
+
+    std::shared_ptr<QuotaAggregator> quota_aggregator_copy = quota_aggregator_;
+
+    check_transport(*quota_request_copy, quota_response,
```
Comment (qiwzhang) [inline src/service_control_client_impl.cc]: quota_transport?

File: src/signature.cc [hunk 2]
```diff
@@ -92,5 +94,38 @@ string GenerateCheckRequestSignature(const CheckRequest& request) {
...
+
+// TODO:: jaebong:: This is the copy of GenerateCheckRequestSignature above.
+// This function might be changed later
+string GenerateAllocateQuotaRequestSignature(const AllocateQuotaRequest& request) {
+  MD5 hasher;
+
+  const QuotaOperation& operation = request.allocate_operation();
+
+	// the value of the operation_id field seems to be different
+  // hasher.Update(operation.operation_id());
```
Comment (qiwzhang) [inline src/signature.cc]: we need operation_name

File: src/quota_aggregator_impl.cc [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+        GOOGLE_LOG(WARNING) << "Last refresh request was not completed yet.";
+      }
+
+      elem->set_is_flushing(true);
+
+      // Setting last check to now to block more check requests to Chemist.
+      elem->set_last_check_time(SimpleCycleTimer::Now());
+
+      // By returning NO_FOUND, caller will send request to server.
+      return Status(Code::NOT_FOUND, "");
```
Comment (qiwzhang) [inline src/quota_aggregator_impl.cc]: This is not good.   We need to call with aggregated metrics.   This is the different between Check() and AllocateQuota()

Comment (qiwzhang) [inline src/quota_aggregator_impl.cc]: One way is to let this item flush out of cache.  The flush call will send the metrics with aggregated values

File: src/quota_aggregator_impl.cc [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+      this, &stack_buffer);
+
+  if (cache_) {
+    string request_signature = GenerateAllocateQuotaRequestSignature(request);
+    QuotaCache::ScopedLookup lookup(cache_.get(), request_signature);
+
+    int64_t now = SimpleCycleTimer::Now();
+
+    // TODO(qiwzhang): supports quota
+    // int scale = GetQuotaScale(request, response);
```
Comment (qiwzhang) [inline src/quota_aggregator_impl.cc]: how do we scale?

**Pull request closed without merging**